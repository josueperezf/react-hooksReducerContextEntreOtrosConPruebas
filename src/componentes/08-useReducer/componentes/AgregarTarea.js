import React from 'react'
import { useForm } from '../../../hooks/useForm';

export const AgregarTarea = ({handleAgregarTarea}) => {
    const [{descripcion}, handleInputChange, reset] = useForm ({
        descripcion: ''
    });
    const handleSubmit = (e)=>{
        e.preventDefault();
        if (descripcion.trim().length ===0) {
            return;
        }
        const nuevaTarea = {
            id:     new Date().getTime(),
            desc:   descripcion,
            done:   false
        };
        handleAgregarTarea(nuevaTarea);
        reset();
    }
    return (
        <>
            <h4>Agregar Tarea</h4>
            <hr/>
            <form onSubmit={handleSubmit} >
                <input
                    type='text'
                    name='descripcion'
                    value={descripcion}
                    className='form-control'
                    placeholder='Aprender...'
                    autoComplete='off'
                    onChange={ handleInputChange }
                />
                <button
                    className='btn btn-outline-primary mt-1 w-100'
                    type='submit'
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
