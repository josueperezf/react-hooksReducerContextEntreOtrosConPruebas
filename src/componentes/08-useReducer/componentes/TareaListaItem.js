import React from 'react'

export const TareaListaItem = ({tarea, i, handleDelete, handleToggle}) => {
    return (
        <li className='list-group-item'  key={tarea.id} >
            <p className={ `${tarea.done && 'complete'}`}
                onClick={()=>handleToggle(tarea.id)}
            >
                {i+1 }. {tarea.desc}
            </p>
            <button
                className='btn btn-danger'
                onClick={()=>handleDelete(tarea.id)}
            >
                Borrar
            </button>
        </li>
    )
}
