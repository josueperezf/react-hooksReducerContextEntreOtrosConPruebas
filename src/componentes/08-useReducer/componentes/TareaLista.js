import React from 'react'
import { TareaListaItem } from './TareaListaItem';
export const TareaLista = ({tareas, handleDelete, handleToggle}) => {
    return (
        <ul className="list-group list-group-flush">
            {
                tareas.map((tarea, i)=>(
                    <TareaListaItem key={tarea.id}  tarea={tarea} i={i} handleDelete={handleDelete} handleToggle={handleToggle} />
                ))
            }
        </ul>
    )
}
