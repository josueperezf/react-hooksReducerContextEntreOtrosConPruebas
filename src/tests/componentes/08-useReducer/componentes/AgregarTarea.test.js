import { shallow } from "enzyme";
import { AgregarTarea } from "../../../../componentes/08-useReducer/componentes/AgregarTarea";

describe('Pruebas en <AgregarTarea/>',()=>{

    const handleAgregarTarea = jest.fn();
    const wrapper = shallow(<AgregarTarea handleAgregarTarea={handleAgregarTarea} />);
    // expect.any es para decir q no me importa el valor solo el tipo de dato que retornara

    test('Debe Mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('no debe de llamar AgregarTarea cuando no hay value en el form', () => {
        const formSubmit = wrapper.find('form').prop('onSubmit');
        // console.log(formSubmit);
        formSubmit({
            preventDefault:()=>{}
        });
        // verifico que haya sido llamado cero veces, hay otra forma de hacer lo mismo pero el quiso variar
        expect(handleAgregarTarea).toHaveBeenCalledTimes(0)
    });

    test('debe de llamar la funcion handleAgregarTarea', () => {
        // con argumento
        const value = 'Aprender Python';
        wrapper.find('input').simulate('change',{
            target:{
                value,
                name:'descripcion'
            }
        });
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({
            preventDefault:()=>{}
        });

        expect(handleAgregarTarea).toHaveBeenCalledTimes(1);
        expect(handleAgregarTarea).toHaveBeenCalledWith( expect.any(Object));
        // id: expect.any(Number), digo que el id no me importa su valor con tal y sea un numero
        expect(handleAgregarTarea).toHaveBeenCalledWith( {
            id: expect.any(Number),
            desc:   value,
            done:   false
        });
    })
    
    
    
});