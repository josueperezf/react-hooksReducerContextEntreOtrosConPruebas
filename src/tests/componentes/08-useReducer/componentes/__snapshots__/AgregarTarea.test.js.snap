// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AgregarTarea/> Debe Mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar Tarea
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="descripcion"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 w-100"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
